libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Shrink.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE GADTs #-}
+
+module Test.Cardano.Ledger.Constrained.Shrink (
+  shrinkEnv,
+) where
+
+import Control.Monad
+import Data.Maybe
+
+import Cardano.Ledger.Core (Era (..))
+import Test.Cardano.Ledger.Constrained.Ast
+import Test.Cardano.Ledger.Constrained.Env
+import Test.Cardano.Ledger.Constrained.Monad
+import Test.Cardano.Ledger.Constrained.Rewrite
+import Test.Cardano.Ledger.Constrained.TypeRep
+
+-- | Shrink an environment subject to the constraints in the given dependency graph.
+--   The strategy is
+--    * pick a variable
+--    * shrink its value to something that still satisfies the defining constraints for the variable
+--    * adjust the values for variables depending on the picked variables to fix constraint
+--      violations resulting from the changed value
+shrinkEnv :: Era era => DependGraph era -> Env era -> [Env era]
+shrinkEnv (DependGraph vs) env =
+  [ env'
+  | (before, (x, cs), after) <- splits vs
+  , env' <- shrinkOneVar env (map fst before) x cs after
+  ]
```

Comment:
Say some thing about how the ordering allows one to know which variables depend upon the variable one picked.

Reply:
👍 Fixed